R/fitNullModel.R

```diff
@@ -193,3 +221,89 @@ nullModelSmall <- function(null.model) {
 isNullModelSmall <- function(null.model) {
     is.null(null.model$cholSigmaInv)
 }
+
+
+.modelString <- function(outcome, covars = NULL, random = NULL, group.var = NULL, inverse_normal = FALSE) {
+    model.outcome <- .modelOutcomeString(outcome, inverse_normal = inverse_normal)
+    model.covars <- if (is.null(covars)) NULL else .modelCovarString(covars)
+    model.random <- if (is.null(random)) NULL else paste(paste0("(1|", random, ")"), collapse=" + ")
+    model.var <- if (is.null(group.var)) NULL else paste0("var(", group.var, ")")
+    model.string <- paste(c(model.covars, model.random, model.var), collapse=" + ")
+    paste(model.outcome, model.string, sep=" ~ ")
+}
+
+.modelCovarString <- function(covars) {
+  paste(covars, collapse=" + ")
+}
+
+.modelOutcomeString <- function(outcome, inverse_normal = FALSE) {
+  if (inverse_normal) {
+    return(sprintf("rankInvNorm(resid(%s))", outcome))
+  } else {
+    return(outcome)
+  }
+}
+
+.setCovMatNames <- function(cov.mat) {
+  if (!is.list(cov.mat)) {
+    cov.mat <- list(cov.mat)
+  }
+  # What about the case where one element of the matrix has names, and the others don't?
+  if (is.null(names(cov.mat))) {
+    names(cov.mat) <- LETTERS[1:length(cov.mat)]
+  } else {
+    if (any(names(cov.mat) == "")) {
+      stop("Some names for cov.mat list are missing.")
+    }
+  }
+  cov.mat
+}
+
+# Update a null model from the old format to the new format.
+.updateNullModelFormat <- function(nullmod) {
+
+  msg <- paste(
+    "This null model was created with an older version of GENESIS and is being updated to use the current version.",
+    "Please consider re-running fitNullModel with the current GENESIS version."
+  )
+
+  # Check if old format or new format.
+  if (!("fit" %in% names(nullmod))) {
+
+    warning(msg)
+
+    # Update.
+    nullmod$fit <- data.frame(
+      outcome = as.vector(nullmod$outcome),
+      workingY = as.vector(nullmod$workingY),
+      fitted.values = as.vector(nullmod$fitted.values),
+      resid.marginal = as.vector(nullmod$resid.marginal),
+      resid.PY = as.vector(nullmod$resid),
+      resid.cholesky = as.vector(nullmod$Ytilde),
+      stringsAsFactors = FALSE
+    )
+    optional_elements <- c("resid.conditional", "sample.id")
+    for (element in optional_elements) {
+      if (element %in% names(nullmod)) {
+        nullmod$fit[[element]] <- nullmod[[element]]
+        nullmod[[element]] <- NULL
+      }
+    }
+    rownames(nullmod$fit) <- rownames(nullmod$model.matrix)
+
+    nullmod$outcome <- NULL
+    nullmod$workingY <- NULL
+    nullmod$fitted.values <- NULL
+    nullmod$resid.marginal <- NULL
+    nullmod$resid <- NULL
+    nullmod$Ytilde <- NULL
+  }
+
+  # Add RSS0
+  if (is.null(nullmod$RSS0)) {
+    warning(msg)
```

Comment:
will we get the same warning message twice? If a model doesn't have `RSS0`, it's also not going to have `fit`. Unless the user did something really weird and modified the output.

Reply:
Yes, right now you get the warning twice. There was a version of the code that has `RSS0` but not the `fit` data frame, so sometimes it has to add both `fit` and `RSS0` and other times it just has to add `RSS0`. Maybe it's not worth printing the warning for missing `RSS0`. Thoughts?

Reply:
No strong opinion. It's fine to get the warning twice, or not bother for the RSS0 change, since it's such a minor change. Or check for either condition and then print the warning once.

Reply:
I ended up taking out the warning in the RSS0 check.